CIP-hardforkSafeguard/README.md

```diff
@@ -0,0 +1,193 @@
+---
+CIP: ?
+Title: Hardfork Safeguard via Stake Representation
+Authors: Jared Corduan <jared.corduan@iohk.io>
+Status: Draft
+Type: Standards Track
+Created: 2022-06-30
+License: CC-BY-4.0
+---
+
+## Simple Summary / Abstract
+
+This CIP replaces a manual safety check regarding the readiness of the network
+for a hardfork with an automatic check.
+
+Ever since the Shelley ledger era, block headers have included a protocol
+version indicating the maximum supported protocol version that the block
+producer is capable of supporting (see section 13, Software Updates, of the
+[Shelley ledger specification](https://hydra.iohk.io/job/Cardano/cardano-ledger/shelleyLedgerSpec/latest/download-by-type/doc-pdf/ledger-spec)).
+This provides a helpful metric for determining how many blocks will be produced
+after a hardfork, since nodes that have not upgraded will no longer produce blocks.
+(Nodes that have not upgraded will fail the `chainChecks` check from Figure 74
+of the Shelley ledger specification, since the major protocol version in the
+ledger state will exceed the node's max major protocol version value,
+and hence can no longer make blocks.)
+
+If most of the blocks in the recent past (e.g. the last twenty-four hours) are
+broadcasting their readiness for a hardfork,
+we know that it is safe to propose an update to the major protocol version
+which triggers a hardfork.
+
+We propose automating this process.
+The ledger state will determine the stake
+(represented as the proportion of the active stake) of all the block producers
+whose last block contained the next major protocol version.
+Moreover, a new protocol parameter `hfThreshold` will be used to reject any
+protocol parameter update that proposes to change the major protocol version
+but does not have enough backing stake.
+
+## Motivation / History
+
+Currently, the governance key holders collectively agree to increase the major
+protocol version.  This allows them to make a human judgment as to the
+readiness of the network for a hardfork (using the mechanism described above).
+Since only a few, well-aligned parties are involved, this is currently easy to
+coordinate.  As we move into the Voltaire phase, where the governance of the
+network is decentralized,
+it is imperative that we codify this human judgment in the protocol itself.
+
+## Specification
+
+### New Protocol Parameter
+
+There will be a new protocol parameter named `hfThreshhold`,
+containing a rational number between zero and one.
+
+### Tracking Hardfork endorsements
+
+The ledger state will maintain a set of stake pool IDs corresponding to the
```

Comment:
One problem this doesn't address is what if you have competing proposals? The scheme described in this CIP only allows for sequential votes. The way this CIP is structure, I don't think this even allows you to do one vote after another because the current structure doesn't have an expiry on upgrade proposals (other than maybe skipping version numbers if a version is deemed to have failed)

Reply:
There are strict rules on how the protocol version can be increased:

https://github.com/input-output-hk/cardano-ledger/blob/7e2f674d2a2d14752d4c2d5abf60b26ae015b9e2/eras/shelley/impl/src/Cardano/Ledger/Shelley/PParams.hs#L519-L520

```
(m + 1, 0) == (m', n') || (m, n + 1) == (m', n')
```

Either the major is increased by exactly one (and the minor reset to zero), or the minor is increased by exactly one (and the major remains unchanged).

Moreover, this proposal is only putting in a safeguard for hardforks (major number increase). So it is always clear what the broadcasted protocol version in the block header is referring to.

Maybe this is only clear if I also explain how the existing protocol parameter update system works? During the voting window, each goverance key can propose a change (they can submit multiple proposals, but the latest one overrides the previous) for the end of the current epoch. If quorum is met, the change happens, otherwise nothing happens and the voting state resets. After the voting window, the each goverance key can stage a vote for the next epoch, which behaves exactly as though they waited until the next epoch and placed a vote during the next voting window.

---

> current structure doesn't have an expiry on upgrade proposals

I think I've explained this above as well. The current structure has harsh expirations. or am I misunderstanding what you meant?

Reply:
I've added a sentence to the end of this paragraph, let me know if it's clear.

Reply:
I don't think you actually addressed the real question - and this is related to my comment above.
Who gets to make such proposals. Let's say entity A wants to make a change to parameter X and therefore submits a proposal to change the protocol version to (m+1, 0) - and entity B wants to leave parameter X unchanged, but wants to change parameter Y instead. The also submit a proposal. Which protocol version would be assigned to that? How would a determination be made, which of the changes to proceed with? How would the SPOs indicate which of the proposals they endorse?

Reply:
> Let's say entity A wants to make a change to parameter X and therefore submits a proposal to change the protocol version to (m+1, 0) - and entity B wants to leave parameter X unchanged, but wants to change parameter Y instead. The also submit a proposal. Which protocol version would be assigned to that?

I guess there are two issues here:
* how the protocol parameter update system works
* how stake pool operators endorse a hard fork

The answer to the first question is:

Whether or not the governance system will move to change the protocol version to  `(m+1, 0)` depends not just on A and B, but also on the other five governance entities (the quorum is 5 of 7 on mainnet). The current system is very basic: at least five of the keys must agree on the entire set of changes. So if entities A - G all want to change Y to 42, but only four of the entities want to change the protocol version to `(m+1, 0)`, nothing is changed at all, not even Y.

The answer to the second question is:

Regardless of what the governance body is doing, if you are a stake pool operator and you are aware of a software update that prepares a hard fork, let's say introducing protocol version to  `(m+1, 0)`, you can:
* signal your willingness to enact the harfork by placing `m+1` in your block headers (the new software will actually do this for you)
* signal your unwillingness to enact the harfork by placing `m` in your block headers (the old software will actually do this for you)

If not enough stake is backed by the bolck producers posting `m+1`, no update proposal can occur which changes the major version to `m+1`, even if quorum is met and even if other protocol parameters we also slated to change.

---

I can try to summarize this in the CIP, since clearly it's still not clear.

Reply:
I've tried again to make this point more clear. Let me know if it is still murky.

Reply:
What happens if 20% of SPO's upgrade, then a bug is found.  50% of the original upgraded SPO then install the new fixed major release and enough other SPO's do as well so the hard fork is successful.  But 10% of the SPO's have the buggy prefix running but it has the same version number?

Reply:
That's an excellent point @WarriorField , and I'm embarrassed I did not think to address it in this CIP. It's far from just a theoretical concern, this came up quite recently, and this is what we did:

https://github.com/input-output-hk/cardano-node/blame/8832f86728ef6a425452b44f2f269acde149448c/cardano-node/src/Cardano/Node/Protocol/Cardano.hs#L201-L206

We fiddled with the minor version. And of course this only worked since the process is still manual. This CIP should address this, thank you!